library/src/main/java/bio/terra/landingzone/stairway/flight/create/reference/resource/step/BaseReferencedResourceStep.java

```diff
@@ -0,0 +1,275 @@
+
+  public static final String REFERENCED_RESOURCE_ID = "referencedResourceId";
+
+  protected static final String FAILED_TO_CREATE_REF_RESOURCE =
+      "Failed to create landing zone {} referenced resource. landingZoneId={}. Error: {}";
+
+  protected static final String RESOURCE_CREATED =
+      "{} resource id='{}' in resource group '{}' successfully created.";
+  private static final String REFERENCED_RESOURCE_NOT_FOUND = "{} doesn't exist.";
+  private static final String FAILED_ATTEMPT_TO_REMOVE_LZ_TAGS =
+      "Failed attempt to remove the landing zone tags for resource: {}. Id={}";
+
+  protected final ArmManagers armManagers;
+
+  protected BaseReferencedResourceStep(ArmManagers armManagers) {
+    this.armManagers = armManagers;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+
+    var landingZoneId = getLandingZoneId(context);
+
+    var azureLandingZoneRequest =
+        getParameterOrThrow(
+            context.getInputParameters(),
+            LandingZoneFlightMapKeys.LANDING_ZONE_CREATE_PARAMS,
+            LandingZoneRequest.class);
+
+    try {
+      var stepDuration =
+          MetricUtils.configureTimerForLzStepDuration(
+              azureLandingZoneRequest.definition(), getArmResourceType().toString());
+      var start = Instant.now().toEpochMilli();
+
+      tagReferencedResourceAndSetContext(context);
+
+      var finish = Instant.now().toEpochMilli();
+      stepDuration.record(Duration.ofMillis(finish - start));
+    } catch (ManagementException e) {
+      return handleManagementException(
+          e, landingZoneId.toString(), azureLandingZoneRequest.definition());
+    } catch (RuntimeException e) {
+      MetricUtils.incrementLandingZoneCreationFailure(azureLandingZoneRequest.definition());
+      throw maybeThrowAzureInterruptedException(e);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  private UUID getLandingZoneId(FlightContext context) {
+    return getParameterOrThrow(
+        context.getInputParameters(), LandingZoneFlightMapKeys.LANDING_ZONE_ID, UUID.class);
+  }
+
+  private void tagReferencedResourceAndSetContext(FlightContext context) {
+    GenericResource resource =
+        findReferencedResourceByArmResourceType(context)
+            .orElseThrow(
+                () ->
+                    new RuntimeException(
+                        String.format(
+                            "A resource of type:%s could not be found in resource group: %s",
+                            getArmResourceType(), getMRGName(context))));
+
+    setLandingZoneResourceTags(context, resource);
+
+    context.getWorkingMap().put(REFERENCED_RESOURCE_ID, resource.id());
+  }
+
+  private void setLandingZoneResourceTags(FlightContext context, GenericResource genericResource) {
+
+    if (genericResource.tags().containsKey(LandingZoneTagKeys.LANDING_ZONE_ID.toString())) {
```

Comment:
Good point. I've made a change based on this. Now when the id is the same as requested, it will not error out and will set all tags again, ensuring idempotency.